### PR TITLE
[mlir][OpenMP] - MLIR to LLVMIR translation support for delayed privatization of allocatables in `omp.target` ops

### PR DESCRIPTION
**Pull request by ergawy**
This PR adds support to translate the `private` clause from MLIR to LLVMIR when used on allocatables in the context of an `omp.target` op.

This replaces https://github.com/llvm/llvm-project/pull/113208.

Parent PR: https://github.com/llvm/llvm-project/pull/116770. Only the latest commit is relevant to the PR.

### CODE REVIEW

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (tblah) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:3828]:
```suggestion
```
nit

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (tblah) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:3942]: How does this stay consistent if we hit `continue` on line 3922?

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (Dinistro) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:3928]: `llvm::` should not be necessary, these are re-exported in the `mlir::` namespace.

**Review by Dinistro — COMMENTED** [review]
Dropped a bunch of nit comments but I cannot do a deep review as I'm not really involved in flang or the OpenMP codegen.

**Comment by ergawy** [comment]
I have opened https://github.com/llvm/llvm-project/pull/116770 as precursor for this one to make matching private variables with their map info ops easier. Please take a look at https://github.com/llvm/llvm-project/pull/116770 first.

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (tblah) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:3924]: I would add a helper function to the MLIR operation. Probably going so far as something like
```
Value getMappedValueForPrivateVar(Value);
```

Or if the O(n) lookup of the variable feels too wasteful, at least something like
```
Value getMappedValueForPrivateVar(unsigned i);
```

Some of the other helpers for private block arguments have helped readability in my opinion e.g. "getCopyMoldArg" and "getDeallocMoldArg".

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (ergawy) [reply mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp]: Done in https://github.com/llvm/llvm-project/pull/116770.

File: mlir/test/Target/LLVMIR/openmp-target-private-allocatable.mlir
Comment (ergawy) [reply mlir/test/Target/LLVMIR/openmp-target-private-allocatable.mlir:1]: Reduced the introduced tests quite a bit.

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (ergawy) [reply mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:3826]: The subscript notation is used in similar situations. I will stick to it if you don't mind.

**Comment by ergawy** [comment]
With https://github.com/llvm/llvm-project/pull/116770 merged, we can go back to this PR. Please take a look when you have time 🙏.

**Review by tblah — APPROVED** [review]
LGTM

**Comment by ergawy** [comment]
Thanks @tblah for the approval. @other reviewers, please let me know if you have any objections to merging this PR.

**Comment by Dinistro** [comment]
> Thanks @tblah for the approval. @other reviewers, please let me know if you have any objections to merging this PR.

I'm fine with this, but I'm not too familiar with these parts to provide an in-depth review.

File: mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td
Comment (NimishMishra) [inline mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td:139]: `privatizerNeedsMap` or `needsMap`?

File: mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td
Comment (NimishMishra) [inline mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td:142]: "descriptor is use in privatization" -> "descriptor is used in privatization"

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp
Comment (tblah) [inline llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:6814]: Does clang/can clang use OMPIRBuilder for `target`? This will change their generated code. Maybe somebody who contributes to clang should review.

**Review by tblah — COMMENTED** [review]
I really like the extra code sharing here. +1 from me but please consider if you need review from other users of OMPIRBuilder.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp
Comment (ergawy) [reply llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:6814]: So far, it is only referenced from the MLIR translation layer. However, I added @ddpagan from the clang side.

Comment (tblah) [reply llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:6814]: Thanks!

File: mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td
Comment (skatrak) [inline mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td:145]: Nit: This can be slightly simplified.
```suggestion
      return !getAllocMoldArg().use_empty();
```

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (skatrak) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp]: I think it's best to early exit to reduce nesting in almost the whole function (https://llvm.org/docs/CodingStandards.html#early-exits).

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (skatrak) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp]:
```suggestion
    if (privVarType == blockArgType)
```

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (skatrak) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:3876]: Ultra-nit: There are several comment lines broken too short.

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (skatrak) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:3877]: Is it possible to make the mapping between values rather than index? According to the example above: `{%6#0, %arg0}`, `{%6#0, %10}` or `{%6#0, (%10, %arg0)}` depending on what's needed. I'm suggesting this because it would be very easy for this to break as soon as some new entry block argument is added between the time this map is created and used.

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (skatrak) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:3915]: Nit: Add short message.

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (skatrak) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp]: Nit: I think using the named accessor helps keep things self-documented as much as possible. At least personally, I'd prefer to keep the logic of what argument of what region represents a certain element to the operation itself, rather than exposing this to users.
```suggestion
        moduleTranslation.mapValue(
            privatizer.getAllocMoldArg(),
            findHostAssociatedValue(privVar, targetOp, mappedPrivateVars,
                                    builder, moduleTranslation));
```

File: mlir/test/Target/LLVMIR/openmp-target-private.mlir
Comment (skatrak) [inline mlir/test/Target/LLVMIR/openmp-target-private.mlir:103]:
```suggestion
// we map the descriptor using the map clause of the omp.target op. Map clauses
```

**Review by skatrak — COMMENTED** [review]
Thank you Kareem, it generally looks good to me. Just a couple of comments.

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (ergawy) [reply mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:3877]: Tried that, and reported my findings here: https://github.com/llvm/llvm-project/pull/116770#issuecomment-2490333779.

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (ergawy) [reply mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp]: I ended up refactoring the code to reuse `allocatePrivateVars`, so this is removed.

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (skatrak) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:1300]: Nit: Feel free to ignore if you disagree or if it introduces implementation issues, but since both things are expected to be passed in the same situation, how about wrapping them in an `std::optional` instead of passing pointers?
```suggestion
                    std::optional<std::pair<omp::TargetOp, llvm::DenseMap<Value, int> &>> targetInfo) {
```
I wouldn't mind introducing a small struct to hold this information in place of the `std::pair` either.

However, if you agree to change `mappedPrivateVars` to `llvm::DenseMap<Value, Value>`, as discussed in another comment in this file, then the suggestion here would be to just pass a pointer to that map. There would be no need to pass `targetOp` in that case. Instead, `blockArg` could just be obtained by querying `(*mappedPrivateVars)[privateVar]`.

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (skatrak) [reply mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:3879]: My comment was more about the implementation of this map specifically, rather than the design of the operation. `mappedPrivateVars` will only contain indices that point to valid block arguments of the operation, so it should be possible to insert the actual value instead. Then, within `findAssociatedValue()`, you wouldn't have to call `targetOp.getRegion().getArgument()` or even pass `targetOp` at all because it would be done here already.

I'm adding a suggestion block below to illustrate what I mean.

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (skatrak) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:3937]:
```suggestion
      mappedPrivateVars.insert({privVar, targetRegion.getArgument(argIface.getMapBlockArgsStart() + (*privateMapIndices)[privVarIdx]}));
```

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (skatrak) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:3985]: Nit: This seems to be the preferred form here, but feel free to leave it unchanged.
```suggestion
    if (failed(handleError(afterAllocas, *targetOp)))
```

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (ergawy) [reply mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp]: I see what you mean. Yup, good suggestion, thank you! Done!

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (ergawy) [reply mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp]: I took your other suggestion.

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (skatrak) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:3874]:
```suggestion
  // privatizer @box.privatizer on the device side. Here we'd record {%6#0, %arg0}
```

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (skatrak) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:3932]:
```suggestion
      // Record the block argument corresponding to this mapvar.
```

**Review by skatrak — APPROVED** [review]
Thank you, LGTM!

**Pull request merged**
